### ISSUE TITLE
Support Marshmallow (6.0)

### ISSUE THREAD

**Issue by mfonville** [opengapps/opengapps#93]
This bug will be the central tracking bug for marshmallow support.
-  [x] The website will have to be updated (initially greyed out, but JavaScript can already be added)
-  [x] Makefile has to be updated
-  [x] build_gapps.sh needs to be updated
-  [x] apk lib extraction has to be skipped for 23+
-  [x] 6.0 packages need to be added to the repos
-  [x] buildbot has to be updated (will be done when aosp rom will be available)

**Comment by bgiesing** [comment]
From TKs thread, do we need to re-evaluate for M? 

----

[quote name="Spasticdroid" post=62385523]A little OT here, but having a look inside the image of Marshmallow Preview 3, it seems (for just about all Google apps) that you won't have to pull out the libraries and put them in their own folders anymore, which saves quite a bit of space, especially with Chrome and WebView :victory:[/QUOTE]Yep. This will also make it easy peasy for anyone to integrate an update from the Play Store right back into their system partition. The value in having an updated GApps package will be greatly diminished if things stay this way through the release. Anyone can sideload an update and integrate right into their system partition without having to wait for an updated GApps package. 

I'm extremely happy to see this change.

**Comment by mfonville** [comment]
We can pursue such a project, to move Google apps from data to system.
But after some consideration, I doubt it will be the holy grail. Since not all gapps are aways available from the play store eligible for installation on all phones (e.g. Google keyboard).
Also installing all apps on 1 go after installation of ROM is convenient. And when updating the ROM, all apps (and their data) would be wiped, unless some addon.d hack would be applied.
Also, symlinks to systemwide libs, that really emulate the Nexus filesystem layout, would not be easily possible.

So I think the 'regular' packages will still be useful on 6.0.
The libraries change is trivial, and will make things more simple. It needs more study for how multi-arch apks will be handled.

**Comment by bgiesing** [comment]
@mfonville What about [this init.d script by @osm0sis](http://forum.xda-developers.com/attachment.php?attachmentid=3431486&d=1438961312). We could include it in the zip?

**Comment by mfonville** [comment]
We could do something like that. But it would need much more work. Like I said earlier, for addon.d stuff. But that script e.g. does not support multi-arch, and would also greatly fail when installing/updating Chrome (and its crazy libs).
But focus should be to first get regular gapps working. After that we can study this.

**Comment by mfonville** [comment]
I studied now the current Marshmallow preview. But apparently not all APKs have their libs fully embedded. The symlinks from e.g. lib/arm/*.so to /system/lib/ still exist for Chrome, Google Keyboard etc.
So it looks like Google has created a Frankenstein hybrid :-/

**Comment by mfonville** [comment]
If the libs don't have to be pre-extracted btw, we can *could* include different versions of each APK in the package. @rapperskull what is your opinion about this? Should we only include the latest version, irrespective of the DPI, or each DPI that we have available even if the version is older, or some kind of hybrid, where this choice is made per application?
This issue/choice is btw related to #97

**Comment by rapperskull** [comment]
As for as the Marshmallow changes, let's wait for the final release to see if there will be other changes. Apps like WebView probably need to have symlinked libs, so that they can be accessed system-wide, so we should make a list of what apps have their libraries extracted and what haven't to correctly build our GApps.
As for as the DPI question, we have to keep in mind that it's just a temporary problem, since it only occurs when the apps get first updated and solves itself when we obtain all the APKs. We can think about including old replacing versions, but we have to find a way to implement the change. Google could for example completely remove a DPI variation and we would still install an old version as a placeholder.
I think that the best is to always include only the latest version (that's why the users always flash the latest package), but there are points to support the other view too.

**Comment by TheCrazyLex** [comment]
Hey, 
I can help you to create some ROM side tweaks :) 

I have a commit ready to auto grant needed permissions for core google apps already. 

Maybe I can help you you on the prebuilt library thing too ^^


Alex

**Comment by rapperskull** [comment]
Hi,
feel free to make a pull request with the changes. Contributors are always welcome.

**Comment by TheCrazyLex** [comment]
Hi, 
@rapperskull but my patch is on ROM level, it applies on frameworks_base.

**Comment by rapperskull** [comment]
OK, then you'll have to share it with ROM devs. We are currently looking for a way to grant the permissions without any change from the ROM side, but if this patch will become mainstream all ROMs will support it and everything will be easier.

**Comment by TheCrazyLex** [comment]
Maybe if you won't find a way, you can create that repository for reference to ROM devs, so they can directly cherry pick from there? 


Also it would be cool if you could share the list of apps to which I should grant permissions ^^


Alex

**Comment by rapperskull** [comment]
All apps from the core package need to have granted permissions by default, and also all the apps that cannot be installed from the Play Store, like Google Play Store itself, Google Calendar Sync, Face Unlock, Google Exchange Services, Android for Work, GCS, Google Play Services, Google NFC Tags, Sound Search (since the Play Store version is outdated), Google Webview and DMAgent.

**Comment by mfonville** [comment]
@TheCrazyLex Thanks for your help.
@rapperskull he did respond to the issue because of the blogpost I wrote: https://plus.google.com/+OpengappsOrg/posts/CmfvzqwQZM7

So the best would be a double approach, both a script in our package, trying to fix the permissions.
But fixing it ROM side first, and making that patch popular, is even more important, because then we'd have a reference for what the script should achieve, and we would be able to test.

The full list of apps that need the permission are those of the list of gappscore that can be found in inc.buildtarget.sh and inc.compatibility.sh

**Comment by TheCrazyLex** [comment]
@rapperskull @mfonville Ok, thank you, very good list :+1: 

And I should really grant ALL permissions to these apps?

**Comment by TheCrazyLex** [comment]
Also I don't really understand the logic of the issue with the prevuilt libraries... I would like to fix, but need to understand first.

**Comment by deadman96385** [comment]
@TheCrazyLex I was wondering if you could upload your changes to enable perms on google apps by default i am trying to make a basic gapps package for an x86 device so i can start getting sdk 23 app versions for open gapps and other users

**Comment by TheCrazyLex** [comment]
@deadman96385 Yes, I am going to upload the patch to the Exodus Gerrit when ready (during the next 2 days) 
I will also need people to test,cause I am not sure if my mechanics can auto allow in time... 


Will update you here :)


Alex

**Comment by TheCrazyLex** [comment]
@mfonville

**Comment by TheCrazyLex** [comment]
Ok, guys.
I am now at a point where I can start adding all the apps and specifying the permissions.

However I have no 6.0 device here... And on 5.1 the apps need nearly every permission.


Maybe someone with 6.0 can look at all the core gapps (also hidden ones) and tell me which permissions they really need on 6.0

I need your help.

Thanks :)


Alex

**Comment by TheCrazyLex** [comment]
Ok, if you'll need it.
here is my document where you can see which permissions are required in 6.0 by which apps

https://docs.google.com/document/d/1OR8c1PkVjHT8OEXZvQ92qnOKMVwKoHz4fzB2wqEqVbk/edit?usp=sharing


Alex

**Comment by TheCrazyLex** [comment]
Ok, my change is ready and online.

@mfonville @rapperskull please review and test it :)

http://exodus-developers.net:8000/#/c/1974/1


Alex

**Comment by rapperskull** [comment]
I'm not a Java expert, but it looks good. I'm currently not able to test. The only doubt is about the permissions. Are you sure you included all of them? How have you compiled the list?

**Comment by TheCrazyLex** [comment]
@rapperskull I am pretty sure the permissions are right. What I did is, I found someone from the Exodus community who has a nexus device with a stock google  6.0 image installed and handed him the list. 

I asked him to find every app from the list and send me a Screenshot of the permissions used. 

Then I wrote them into the list and... Then wrote the code using Google's  own autogrant system. 

I cannot test myself though as my phone is still not booting with 6.0..

If something is missing a tester would see it in the log and post it here :) 


Alex

**Comment by mfonville** [comment]
@TheCrazyLex thanks for your work. Unfortunately I have no 6.0 device myself, so we are dependent on when @rapperskull will have time to do so.

**Comment by deadman96385** [comment]
I have a work in 6.0 Rom just no 23 sdk apps for it because it being x86

**Comment by TheCrazyLex** [comment]
@mfonville maybe I can help to resolve the other 6.0 issues too? Just tell me what to do :)

**Comment by gladiac** [comment]
Only granting permission by the package name open a security hole. You could install an app with the same name and get the same privileges as the gapps!

You should also check the signing key of the google app!

**Comment by gladiac** [comment]
However,  how is this done on a stock ROM? Did someone check the framework code?

**Comment by TheCrazyLex** [comment]
@gladiac well I've done it in the style Google does it with the stock phone app etc. 

Also it is not a security issue as Android for example automatically grants preset permissions to any browser. So if my patch is a security issue then Android has a big security issue too. 

Also the user decides by himself what he installs from unknown sources and what he doesn't. 
Play Store is 100% safe in this case

I only grant critical permissions to Google play services, every Gapps package has that included. 


Alex

**Comment by TheCrazyLex** [comment]
And all these packages are core packages, means included in every Gapps install

**Comment by mfonville** [comment]
@TheCrazyLex 
One of the other "big" issues that is still open, is that Google's ROMs apparently accept APKs where the libs are still inside the apk, and they are loaded/extracted on-demand. While the AOSP ROMs still expect the libs to be pre-extracted. If you could look into that issue? What causes the difference in behaviour? If that would be fixed, we would be good to go!

**Comment by TheCrazyLex** [comment]
@mfonville I will look into it :)

You are saying that it was working in 5.1.

I mean this extracting on-demand

Alex

**Comment by mfonville** [comment]
No, in 5.1 it wasn't working, 5.1 needed the pre-extracted libs. But that was on both aosp and Google ROMs the same.
But now, Google ROMs seem to be able to do it on-demand, but aosp not.
So Google is doing a trick that ain't in aosp (enabled?) (Yet?)

**Comment by TheCrazyLex** [comment]
@mfonville it is normal, that Google doesn't publish all code....

Most probably we have no chance :(
I will however recheck

Also I wouldn't say it's a deal breaker/blocker issue as you had to
pre-extract libs in 5.1 too.

What do you think?

Alex

**Comment by mfonville** [comment]
It is not a deal breaking or blocking issue per se, *but* since on Marshmallow all APKs need to be fully intact (and thus the libs are not allowed to be removed) it takes much more space on /system/ if the libs are stored twice, once within the apk, and once pre-extracted.

Also it negatively influences our "For Stock" flashing method, that does support flashing on stock ROMs. (Since that one would then expect different behaviour).

I am sure Google must have included it in the code somewhere, since it is a big change to the core of the basesystem. It is very hard to believe this change to be propietary.

We could also look how other vendors do this on their 6.0 images.

**Comment by deadman96385** [comment]
I think i have found the code that allows libraries to be in apk's as well as a way to enable it on aosp builds i am build a build now with my changes i will let you know on the results

**Comment by TheCrazyLex** [comment]
@deadman96385 Awesome man! :)

Please share the code so I can take a look :)

I think the permission issue is closed now, the code is stable and working.

**Comment by deadman96385** [comment]
Here is the commit for it, i am currently making a test build with the default set to false https://github.com/CyanogenMod/android_frameworks_base/commit/ff193d642eea7128faad837d19e347cd25212c27

**Comment by gladiac** [comment]
Why not check the signature of the apk in additon? I increases the security and isn't hard to do ...

**Comment by mfonville** [comment]
I got an extra update for what @deadman96385 did find.
That method requires that the libs are not deflated, but 'stored' in the APK.
Something that is done with the APKs on the nexus image, but not with APKS from the Play Store.
So I will add something to the build script to fix that.

**Comment by TheCrazyLex** [comment]
@mfonville @deadman96385 Good work guys :)

@gladiac I will look into it.

**Comment by mfonville** [comment]
This will make sure that the libs are stored uncompressed within the APK, when building for Marshmallow: https://github.com/opengapps/opengapps/commit/90b34ea1d91aa35ebe1853fb6f74cfdb4489889f

**Comment by TheCrazyLex** [comment]
@mfonville does that mean your commit in combination with the cm commit resolves the issue?

**Comment by deadman96385** [comment]
The commit is a aosp commit and his new commit fixes it so that no change is need to the rom for it work i have just tested it.

**Comment by TheCrazyLex** [comment]
@deadman96385 So that means we are good to go and this issue can be marked as resolved :) 

Alex

**Comment by deadman96385** [comment]
Yep you are corrected

**Comment by TheCrazyLex** [comment]
Ok, last thing to do which isn't critical would be add more security by verfifying the signature. 

Maybe @gladiac can help, as he said it is easy to do so :)

**Comment by mfonville** [comment]
Thank you all for your hard work :-) and good collaboration between all parties!
The bug can finally be closed

**Comment by mfonville** [comment]
@TheCrazyLex 

Hey Alex,

could you maybe make for us a beta patch that also adds the following permissions for the Google Dialer (com.android.dialer)? Because we would like to try if the Google Dialer would work on Stock ROMs, but we think that the lack of permissions is obstructing it to work, and we would like to try it:

   <uses-permission android:name="android.permission.ACCESS_WIFI_STATE"/>
   <uses-permission android:name="android.permission.CALL_PHONE"/>
   <uses-permission android:name="android.permission.READ_CONTACTS"/>
   <uses-permission android:name="android.permission.WRITE_CONTACTS"/>
   <uses-permission android:name="android.permission.READ_CALL_LOG"/>
   <uses-permission android:name="android.permission.WRITE_CALL_LOG"/>
   <uses-permission android:name="android.permission.READ_PROFILE"/>
   <uses-permission android:name="android.permission.MANAGE_ACCOUNTS"/>
   <uses-permission android:name="android.permission.GET_ACCOUNTS"/>
   <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>
   <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION"/>
   <uses-permission android:name="android.permission.INTERNET"/>
   <uses-permission android:name="android.permission.NFC"/>
   <uses-permission android:name="android.permission.PROCESS_OUTGOING_CALLS"/>
   <uses-permission android:name="android.permission.READ_PHONE_STATE"/>
   <uses-permission android:name="android.permission.MODIFY_AUDIO_SETTINGS"/>
   <uses-permission android:name="android.permission.MODIFY_PHONE_STATE"/>
   <uses-permission android:name="android.permission.WAKE_LOCK"/>
   <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
   <uses-permission android:name="android.permission.WRITE_SETTINGS"/>
   <uses-permission android:name="android.permission.USE_CREDENTIALS"/>
   <uses-permission android:name="android.permission.VIBRATE"/>
   <uses-permission android:name="android.permission.READ_SYNC_SETTINGS"/>
   <uses-permission android:name="com.android.voicemail.permission.ADD_VOICEMAIL"/>
   <uses-permission android:name="com.android.voicemail.permission.WRITE_VOICEMAIL"/>
   <uses-permission android:name="com.android.voicemail.permission.READ_VOICEMAIL"/>
   <uses-permission android:name="android.permission.ALLOW_ANY_CODEC_FOR_PLAYBACK"/>
   <uses-permission android:name="com.android.launcher.permission.INSTALL_SHORTCUT"/>
   <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>
   <uses-permission android:name="com.google.android.providers.gsf.permission.READ_GSERVICES"/>
   <uses-permission android:name="android.permission.BROADCAST_STICKY"/>
   <uses-permission android:name="android.permission.INTERACT_ACROSS_USERS"/>
   <uses-permission android:name="android.permission.STOP_APP_SWITCHES"/>
   <uses-permission android:name="android.permission.CONTROL_INCALL_EXPERIENCE"/>
   <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE"/>


Thanks!

**Comment by TheCrazyLex** [comment]
@mfonville of course :)  I will do that tomorrow as it's pretty late here
for today :)

**Comment by mfonville** [comment]
@TheCrazyLex Thanks!  it is here late too (it is already 'tomorrow' :-P)
No hurries and sleep well! :-)

**Comment by TheCrazyLex** [comment]
@mfonville I have prepared a commit, the thing is that Google already grants permissions to the default dialer app. You can even see it in the debug log.

I am adding some additional perms here: However it needs to be tested hard:
http://exodus-developers.net:8000/#/c/2217/1



Alex

**Comment by TheCrazyLex** [comment]
Also maybe you have a log so I can see what's missing for the dialer? :)

**Comment by mfonville** [comment]
I don't have a logcat, since I don't have Marshmallow myself. So I also have to work with the feedback I get from others :-)
I will pass this commit on to those busy with it and will report back asap

**Comment by mfonville** [comment]
@TheCrazyLex @rapperskull @deadman96385 
Could you guys check if https://github.com/opengapps/opengapps/tree/mmpermissions works for you on a ROM without Alex's patch?

**Comment by Blystad** [comment]
Has anyone tried doing a dump of `framework-res.apk` on a standard system image? Usually Google do changes like these using `config.xml` overlays...My feeling is that they would've done that here too.

Second alternative is a xml file on the system image.

Third alternative: decompile the framework and check if they have a special `DefaultPermissionGrantPolicy.java` (since I think you can overlay java files as well, though I'm not 100% sure).

**Comment by Blystad** [comment]
TL;DR: I reverse engineered the binaries on Google Nexus 6 and from what I can see, Google does not need a permission grant patch like we do.

Checked `Settings.apk` ([which is signed with the platform certificate](https://github.com/android/platform_packages_apps_settings/blob/android-6.0.1_r1/Android.mk)) and compared that to `PrebuiltGmsCore.apk`

Settings.apk certificate:
```
Owner: CN=Android, OU=Android, O=Google Inc., L=Mountain View, ST=California, C=US
Issuer: CN=Android, OU=Android, O=Google Inc., L=Mountain View, ST=California, C=US
Serial number: ba901ad4221ec1d0
Valid from: Thu Aug 14 21:15:30 CEST 2014 until: Mon Dec 30 20:15:30 CET 2041
Certificate fingerprints:
	 MD5:  06:5D:BD:18:EC:E5:C2:CE:99:A9:4B:02:04:E9:DA:DE
	 SHA1: 95:5C:D0:66:54:9A:F5:08:28:9C:6B:10:D1:A0:2F:FF:11:E1:14:EE
	 SHA256: AD:9E:31:40:25:14:E5:4F:25:89:16:BB:6F:5F:42:D6:7F:83:78:24:F4:6D:83:C6:5C:68:BE:8E:63:49:98:F6
	 Signature algorithm name: SHA1withRSA
	 Version: 3
...
```

PrebuiltGmsCore
```
Owner: CN=Android, OU=Android, O=Google Inc., L=Mountain View, ST=California, C=US
Issuer: CN=Android, OU=Android, O=Google Inc., L=Mountain View, ST=California, C=US
Serial number: c2e08746644a308d
Valid from: Fri Aug 22 01:13:34 CEST 2008 until: Tue Jan 08 00:13:34 CET 2036
Certificate fingerprints:
	 MD5:  CD:E9:F6:20:8D:67:2B:54:B1:DA:CC:0B:70:29:F5:EB
	 SHA1: 38:91:8A:45:3D:07:19:93:54:F8:B1:9A:F0:5E:C6:56:2C:ED:57:88
	 SHA256: F0:FD:6C:5B:41:0F:25:CB:25:C3:B5:33:46:C8:97:2F:AE:30:F8:EE:74:11:DF:91:04:80:AD:6B:2D:60:DB:83
	 Signature algorithm name: MD5withRSA
	 Version: 3
...
```

The certificates doesn't match...

Then I decided to do a quick grep:
```
{ _system }  » grep -r "googlequicksearch" .                                                                                                                                                                                                                                                           /tmp/shamu/_system  
grep: ./lost+found: Permission denied
Binary file ./app/Chrome/oat/arm/Chrome.odex matches
Binary file ./app/GoogleEars/oat/arm/GoogleEars.odex matches
Binary file ./app/GoogleHome/oat/arm/GoogleHome.odex matches
Binary file ./app/GoogleJapaneseInput/oat/arm/GoogleJapaneseInput.odex matches
Binary file ./app/LatinImeGoogle/oat/arm/LatinImeGoogle.odex matches
Binary file ./app/Maps/oat/arm/Maps.odex matches
Binary file ./app/Music2/oat/arm/Music2.odex matches
Binary file ./app/WebViewGoogle/oat/arm/WebViewGoogle.odex matches
grep: ./bin/install-recovery.sh: Permission denied
grep: ./bin/qmuxd: Permission denied
grep: ./bin/run-as: Permission denied
grep: ./bin/uncrypt: Permission denied
grep: ./etc/dhcpcd/dhcpcd-run-hooks: Permission denied
Binary file ./framework/arm/boot.oat matches
Binary file ./priv-app/ManagedProvisioning/ManagedProvisioning.apk matches
Binary file ./priv-app/Phonesky/oat/arm/Phonesky.odex matches
Binary file ./priv-app/PrebuiltGmsCore/oat/arm/PrebuiltGmsCore.odex matches
Binary file ./priv-app/TriggerEnroll/oat/arm/TriggerEnroll.odex matches
Binary file ./priv-app/Velvet/Velvet.apk matches
Binary file ./priv-app/Velvet/lib/arm/liboffline_actions_jni.so matches
Binary file ./priv-app/Velvet/oat/arm/Velvet.odex matches
```

I think the interesting one here is:
```
Binary file ./framework/arm/boot.oat matches
```
But I extracted the dex file from the oat file, decompiled that, and that was due to this: https://github.com/android/platform_frameworks_base/blob/android-6.0.1_r1/core/java/android/content/pm/PackageInstaller.java#L395

Then I decided, for sanity, to extract the services.dex and check the Google `DefaultPermissionGrantPolicy.java` and it looks identical to the AOSP one: http://pastebin.com/z6zvKHai (Bear in mind that, when a Java class is compiled, some information is lost, and static variables gets substituted).

My findings therefor seem to indicate that Google does not use a patch like the above...So, that brings the question to - what does Google actually do?

Well, lets start by looking at the patch, it grants permissions to these packages:
* `com.google.android.apps.gcs` - should be covered by [Carrier Provisioning Service](https://android.googlesource.com/platform/frameworks/base/+/marshmallow-release/services/core/java/com/android/server/pm/DefaultPermissionGrantPolicy.java#368) (I think gcs is only used by Project Fi? )
* `com.google.android.syncadapters.contacts` - should be covered by [Contacts provider sync adapters](https://android.googlesource.com/platform/frameworks/base/+/marshmallow-release/services/core/java/com/android/server/pm/DefaultPermissionGrantPolicy.java#419)
* `com.google.android.setupwizard` - should be covered by [SetupWizard](https://android.googlesource.com/platform/frameworks/base/+/marshmallow-release/services/core/java/com/android/server/pm/DefaultPermissionGrantPolicy.java#252)
* `com.android.vending` - falls in under [Verifier](https://android.googlesource.com/platform/frameworks/base/+/marshmallow-release/services/core/java/com/android/server/pm/DefaultPermissionGrantPolicy.java#242)

and these:
* `com.google.android.gms` - this is for the NetworkLocationProvider from what I can see...Which seems to indicate it should be covered by [Location](https://android.googlesource.com/platform/frameworks/base/+/marshmallow-release/services/core/java/com/android/server/pm/DefaultPermissionGrantPolicy.java#542) which gets set if you add `com.google.android.gms` to [LocationProviders]((https://android.googlesource.com/platform/frameworks/base/+/android-6.0.1_r1/services/core/java/com/android/server/LocationManagerService.java#222)) See [cyanogenmod_overlay](https://github.com/CyanogenMod/android_vendor_cm/blob/cm-12.1/overlay/common/frameworks/base/core/res/res/values/config.xml#L18) for an example.
* `com.google.android.googlequicksearchbox` - I think this is covered by [Voice Interaction](https://android.googlesource.com/platform/frameworks/base/+/marshmallow-release/services/core/java/com/android/server/pm/DefaultPermissionGrantPolicy.java#509). However, if there's multiple packages implementing the `android.service.voice.VoiceInteractionService` intent, then the [first one will be chosen](https://github.com/android/platform_frameworks_base/blob/master/services/voiceinteraction/java/com/android/server/voiceinteraction/VoiceInteractionManagerService.java#L345). If the ROM ships with QuickSearchBox (standard AOSP one) or Voice Dialer, then this could result in the wrong app getting these permissions.

and then these, which I leave as an exercise to the reader:
* `com.google.android.backuptransport`
* `com.google.android.gsf`
* `com.google.android.gsf.login`

it should be possible to find a permission grant for them as well. I also think permissions gets shared among the same "shared user id". I.e. if one Google App has the permission, then another Google App can get the same permission without specifying it. (It was at least that way in Lollipop and before)

**Comment by mfonville** [comment]
But how should we now handle this? Remove by default all AOSP/CM apps that claim any conflicting permission? 
E.g. when the ROM does ship with the CM Setupwizard, it is removed by the opengapps install script, but the permissions are still not automatically granted to the Google SetupWizard instead. So it is not per se a "first come - first server" for all apps on a per-app basis.

Also, I tried finally myself the mmpermissions-branch, and it does not matter how I edit i.e. the SetupWizard runtime-permissions from recovery, they are always reset to value "20" when booting CM for the first time.

**Comment by Blystad** [comment]
The commit from TeamExynos is one way of solving these issues. I'm merely saying it should work out of the box for standard AOSP builds.

The code points towards first come first served. The code queries for a given intent and return the first component fulfilling that. Wierd that it doesn't happen on CM.

Best bet is debugging the package manager system service to figure out what goes wrong. I would do that myself, but I'm still stuck in the Lollipop world :)

**Comment by TheCrazyLex** [comment]
Lol Team Exynos xD

**Comment by Blystad** [comment]
*Exodus ;) (I wrote this on a phone and didn't bother checking until someone called me out)

**Comment by mfonville** [comment]
@Blystad thanks for further explanation.
I wonder which 'default' intent should match GMS Play Services in the AOSP source

**Comment by Blystad** [comment]
> * `com.google.android.gms` - this is for the NetworkLocationProvider from what I can see...Which seems to indicate it should be covered by [Location](https://android.googlesource.com/platform/frameworks/base/+/marshmallow-release/services/core/java/com/android/server/pm/DefaultPermissionGrantPolicy.java#542) which gets set if you add `com.google.android.gms` to [LocationProviders]((https://android.googlesource.com/platform/frameworks/base/+/android-6.0.1_r1/services/core/java/com/android/server/LocationManagerService.java#222)) See [cyanogenmod_overlay](https://github.com/CyanogenMod/android_vendor_cm/blob/cm-12.1/overlay/common/frameworks/base/core/res/res/values/config.xml#L18) for an example.

**Comment by mfonville** [comment]
@Blystad indeed the permissions seem to match. Unfortunately I am still not able to reproduce the automatic granting of permissions to the GApps by removing a lot of the AOSP/CM default apps yet. So I guess I will have to tinker a bit more with those.

We are sure the intents are checked at runtime, right?
